reps/2023-03-10-batch-inference.md

```diff
@@ -227,28 +237,176 @@ def preprocess(batch: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
     batch["image"] = batch["image"].transpose(0, 3, 1, 2)
     return batch
 
-predictor_creator = lambda: MXNetPredictor(gluon.model_zoo.vision.resnet50_v1(pretrained=True))
-dataset.map_batches(preprocess, batch_format="numpy").map_predictor(predictor_creator)
+predictor = MXNetPredictor(gluon.model_zoo.vision.resnet50_v1(pretrained=True))
+dataset.map_batches(preprocess, batch_format="numpy").map_batches(predictor_creator, compute=ray.data.ActorPoolStrategy(size=8))
 ```
 
+## Internal interfaces & implementation details
+1. Introduce a DeveloperAPI `BatchMappable` interface that `Preprocessor` and `Predictor` subclass. This interface will define the following:
+- Logic for the actual computation (either prediction or preprocessing)
+- Validating the user provided batch_format
+- Validating the user provided compute strategy
 
+```python
+@DeveloperAPI
+class BatchMappable(Callable[[T], U]):
+   """Subclass this if you want your class to be passable to Dataset.map_batches."""
+   
+    def __call__(self, batch: T) -> U:
+        """Transform the provided batch.
+
+        Args:
+            batch: The batch to transform.
+        """
+        
+        raise NotImplementedError
+
+    def setup(self):
+        """Additional setup logic to run.
+
+        If a callable class is being used, this function will be run in the `__init__`
+        of the Callable class.
+        """
+        pass
+       
+    def validate_batch_format(self, batch_format: Optional[str]) -> Optional[str]:
+        """Validate that the provided `batch_format` is compatible with this `BatchMappable`.
+
+        Raises an error if the provided `batch_format` is not supported with this `BatchMappable` 
+        (for example Pyarrow format for Predictors.)
+
+        If `batch_format` is "default", returns the preferred `batch_format` for this `BatchMappable`.
+
+        Args:
+            batch_format: The `batch_format` that is provided to `map_batches`.
+
+        Returns:
+            The batch format to use.
+        """
+       return None
+
+    def validate_compute_strategy(self, strategy):
+        """Validate that the provided `strategy` is compatible with this `BatchMappable`.
+
+        For example, enforce that `ActorPoolStrategy` is provided for Predictors.
+        """
+        pass
+
+    def use_callable_class(self) -> bool:
+        """Indicates whether this `BatchMappable` should use a callable class."""
```

Comment:
Should this always be true?

Reply:
Idts, for Preprocessors we don't want to use callable classes.

But for Predictors, yes, should always be True

Reply:
How would that work?

Reply:
We could have a validator like

```
def validate_args(self, **kwargs):
   pass
```

And Datasets could pass in a kwargs dict like `{"compute": ..., "batch_size": ..., ...}`.